File: apps/api/src/jobs/email.queue.ts
import { Queue, Worker, type JobsOptions } from 'bullmq';
import { Prisma } from '@prisma/client';
import { getRedis } from '../lib/redis.js';
import { sendMail } from '../lib/mailer.js';
import { bookingCancelledEmail, bookingConfirmedEmail, waitlistOfferEmail } from '../lib/emails.js';
import { offerUrl, renderQrDataUrl, verifyUrl } from '../lib/qr.js';
import { prisma } from '../lib/prisma.js';

const QUEUE = 'email';

export type EmailJob =
  | { kind: 'booking-confirmed'; bookingId: string }
  | { kind: 'booking-cancelled'; bookingId: string }
  | { kind: 'waitlist-offer'; entryId: string };

const JOB_OPTIONS: JobsOptions = {
  // Five tries over roughly a minute and a half. A provider blip should not
  // cost somebody their ticket, and a permanent failure should stop quickly.
  attempts: 5,
  backoff: { type: 'exponential', delay: 3_000 },
  removeOnComplete: { count: 50 },
  removeOnFail: { count: 100 },
};

let queue: Queue<EmailJob> | null = null;

function getQueue(): Queue<EmailJob> | null {
  const connection = getRedis();
  if (!connection) return null;
  queue ??= new Queue<EmailJob>(QUEUE, { connection });
  return queue;
}

/**
 * Hands an email to the queue and returns immediately.
 *
 * Never awaited by a request handler in a way that can fail it. A booking is
 * already committed by the time this runs; if the queue is unreachable the
 * booking still stands and the failure is logged loudly rather than turned
 * into a 500 for a customer whose seat is confirmed in the database.
 */
export async function enqueueEmail(job: EmailJob): Promise<void> {
  const subject = job.kind === 'waitlist-offer' ? job.entryId : job.bookingId;
  const q = getQueue();
  if (!q) {
    console.warn(`[email] REDIS_URL not set — ${job.kind} for ${subject} was not queued`);
    return;
  }
  try {
    await q.add(job.kind, job, JOB_OPTIONS);
  } catch (err) {
    console.error(`[email] could not queue ${job.kind} for ${subject}`, err);
  }
}

/**
 * Renders and sends. Reads the booking fresh rather than trusting a payload
 * serialised minutes ago — by the time a retry runs, the booking may have been
 * cancelled, and sending a confirmation for a cancelled booking is worse than
 * sending nothing.
 */
async function process(job: EmailJob) {
  if (job.kind === 'waitlist-offer') return processOffer(job.entryId);

  const booking = await prisma.booking.findUnique({
    where: { id: job.bookingId },
    select: {
      reference: true,
      status: true,
      qrToken: true,
      customer: { select: { name: true, email: true } },
      show: {
        select: {
          startsAt: true,
          event: { select: { title: true, venue: { select: { name: true } } } },
        },
      },
      seats: {
        select: {
          priceAtBooking: true,
          showSeat: { select: { seat: { select: { row: true, number: true } } } },
        },
      },
    },
  });

  if (!booking) {
    console.warn(`[email] booking ${job.bookingId} no longer exists, skipping`);
    return;
  }

  const seats = booking.seats.map((s) => `${s.showSeat.seat.row}${s.showSeat.seat.number}`);

  if (job.kind === 'booking-cancelled') {
    await sendMail({
      to: booking.customer.email,
      subject: `Cancelled — ${booking.reference}`,
      html: bookingCancelledEmail({
        reference: booking.reference,
        eventTitle: booking.show.event.title,
        seats,
      }),
    });
    return;
  }

  if (booking.status !== 'CONFIRMED') {
    console.warn(`[email] booking ${booking.reference} is ${booking.status}, not confirming`);
    return;
  }

  // Decimal arithmetic, not Number(). A float cannot hold 0.10, and money that
  // is off by a cent in an email is money the customer will ask about.
  const total = booking.seats.reduce((sum, s) => sum.add(s.priceAtBooking), new Prisma.Decimal(0));
  const qrDataUrl = await renderQrDataUrl(booking.qrToken);

  await sendMail({
    to: booking.customer.email,
    subject: `Your tickets — ${booking.show.event.title} (${booking.reference})`,
    html: bookingConfirmedEmail({
      name: booking.customer.name,
      reference: booking.reference,
      eventTitle: booking.show.event.title,
      venue: booking.show.event.venue.name,
      startsAt: booking.show.startsAt.toUTCString(),
      seats,
      total: total.toFixed(2),
      qrDataUrl,
      verifyUrl: verifyUrl(booking.qrToken),
    }),
    attachments: [
      {
        filename: `ticket-${booking.reference}.png`,
        content: Buffer.from(qrDataUrl.split(',')[1]!, 'base64'),
      },
    ],
  });

  console.log(`[email] sent ${job.kind} for ${booking.reference} to ${booking.customer.email}`);
}

/**
 * Re-reads the entry rather than trusting the payload. By the time a retry
 * runs, the offer may have expired and moved on to somebody else — emailing a
 * link that is already dead is worse than emailing nothing.
 */
async function processOffer(entryId: string) {
  const entry = await prisma.waitlistEntry.findUnique({
    where: { id: entryId },
    select: {
      status: true,
      offerToken: true,
      offerExpiresAt: true,
      customer: { select: { name: true, email: true } },
      category: { select: { name: true, price: true } },
      show: {
        select: {
          startsAt: true,
          event: { select: { title: true, venue: { select: { name: true } } } },
        },
      },
    },
  });

  if (!entry || entry.status !== 'OFFERED' || !entry.offerToken || !entry.offerExpiresAt) {
    console.warn(`[email] waitlist offer ${entryId} is no longer open, skipping`);
    return;
  }

  const minutes = Math.max(1, Math.round((entry.offerExpiresAt.getTime() - Date.now()) / 60_000));

  await sendMail({
    to: entry.customer.email,
    subject: `A seat opened up — ${entry.show.event.title}`,
    html: waitlistOfferEmail({
      name: entry.customer.name,
      eventTitle: entry.show.event.title,
      venue: entry.show.event.venue.name,
      startsAt: entry.show.startsAt.toUTCString(),
      category: entry.category.name,
      price: entry.category.price.toFixed(2),
      minutes,
      offerUrl: offerUrl(entry.offerToken),
    }),
  });

  console.log(`[email] sent waitlist-offer ${entryId} to ${entry.customer.email}`);
}

let worker: Worker<EmailJob> | null = null;

export function startEmailWorker() {
  const connection = getRedis();
  if (!connection) {
    console.warn('email worker not started — REDIS_URL is not set');
    return async () => {};
  }

  worker = new Worker<EmailJob>(QUEUE, async (job) => process(job.data), {
    connection,
    // The single most important setting on Upstash's metered free tier.
    // Default is 5s, which means a worker with nothing to do issues ~12
    // blocking commands a minute — about 518,000 a month, the entire free
    // allowance, before a single email is sent. At 60s it costs ~43,000.
    // A pushed job wakes the blocked worker immediately, so this is not
    // added latency.
    drainDelay: 60,
    concurrency: 3,
  });

  worker.on('failed', (job, err) =>
    console.error(`[email] job ${job?.id} failed (attempt ${job?.attemptsMade}):`, err.message),
  );

  console.log('email worker running (drainDelay 60s)');
  return async () => {
    await worker?.close();
    worker = null;
  };
}

File: apps/api/src/ticket_api/jobs/email_queue.py
"""
The email queue.

BullMQ has no Python equivalent, so this is ARQ: same shape (Redis-backed,
retries with backoff, a separate worker process), a different library.

Nothing in a request path awaits a send. A booking is confirmed in Postgres
before anything is queued, and a mail provider must never be able to fail a
booking the customer has already made.
"""

from __future__ import annotations

import base64
from datetime import timedelta
from decimal import Decimal
from typing import Any, Literal, TypedDict

from arq import create_pool
from arq.connections import RedisSettings
from sqlalchemy import select

from ..config import IS_TEST, settings
from ..db import Session
from ..lib.emails import (
    booking_cancelled_email,
    booking_confirmed_email,
    show_cancelled_email,
    waitlist_offer_email,
)
from ..lib.mailer import Mail, send_mail
from ..lib.qr import offer_url, render_qr_data_url, verify_url
from ..models import (
    Booking,
    BookingSeat,
    BookingStatus,
    Event,
    Seat,
    SeatCategory,
    Show,
    ShowSeat,
    User,
    Venue,
    WaitlistStatus,
    utcnow,
)

EmailKind = Literal["booking-confirmed", "booking-cancelled", "show-cancelled", "waitlist-offer"]


class EmailJob(TypedDict, total=False):
    kind: EmailKind
    bookingId: str
    entryId: str


def _redis_settings() -> RedisSettings | None:
    if not settings.REDIS_URL:
        return None
    return RedisSettings.from_dsn(settings.REDIS_URL)


_pool: Any = None


async def enqueue_email(job: EmailJob) -> None:
    """
    Hands an email to the queue and returns immediately.

    Never awaited by a request handler in a way that can fail it. If the queue
    is unreachable the booking still stands and the failure is logged loudly
    rather than turned into a 500 for a customer whose seat is already
    confirmed in the database.
    """
    global _pool
    subject = job.get("entryId") or job.get("bookingId")

    # Tests never reach the real queue. REDIS_URL points at the live Upstash
    # instance, and a suite that enqueues thousands of jobs there is the same
    # mistake as one that writes to the production database — plus it made the
    # booking tests fourteen times slower, one network round trip at a time.
    if IS_TEST:
        return

    redis_settings = _redis_settings()
    if redis_settings is None:
        print(f"[email] REDIS_URL not set — {job['kind']} for {subject} was not queued")
        return

    try:
        if _pool is None:
            _pool = await create_pool(redis_settings)
        await _pool.enqueue_job("send_email", job)
    except Exception as err:  # noqa: BLE001 - a queue failure must never fail a booking
        print(f"[email] could not queue {job['kind']} for {subject}: {err}")


def _utc_string(value: Any) -> str:
    """Matches JavaScript's toUTCString(), which is what the old emails showed."""
    return value.strftime("%a, %d %b %Y %H:%M:%S GMT")


async def send_email(_ctx: dict[str, Any], job: EmailJob) -> None:
    """
    Renders and sends.

    Reads the booking fresh rather than trusting a payload serialised minutes
    ago — by the time a retry runs the booking may have been cancelled, and
    sending a confirmation for a cancelled booking is worse than sending
    nothing.
    """
    if job["kind"] == "waitlist-offer":
        await _process_offer(job["entryId"])
        return

    booking_id = job["bookingId"]
    async with Session() as session:
        booking = (
            (await session.execute(select(Booking).where(Booking.id == booking_id)))
            .scalars()
            .first()
        )
        if booking is None:
            print(f"[email] booking {booking_id} no longer exists, skipping")
            return

        customer = (
            (await session.execute(select(User).where(User.id == booking.customer_id)))
            .scalars()
            .one()
        )
        show = (
            (await session.execute(select(Show).where(Show.id == booking.show_id))).scalars().one()
        )
        event = (
            (await session.execute(select(Event).where(Event.id == show.event_id))).scalars().one()
        )
        venue = (
            (await session.execute(select(Venue).where(Venue.id == event.venue_id))).scalars().one()
        )

        rows = (
            await session.execute(
                select(BookingSeat, Seat)
                .join(ShowSeat, ShowSeat.id == BookingSeat.show_seat_id)
                .join(Seat, Seat.id == ShowSeat.seat_id)
                .where(BookingSeat.booking_id == booking_id)
                .order_by(Seat.row.asc(), Seat.number.asc())
            )
        ).all()

    seats = [f"{seat.row}{seat.number}" for _, seat in rows]

    if job["kind"] == "show-cancelled":
        await send_mail(
            Mail(
                to=customer.email,
                subject=f"Cancelled — {event.title}",
                html=show_cancelled_email(
                    reference=booking.reference,
                    event_title=event.title,
                    venue=venue.name,
                    starts_at=_utc_string(show.starts_at),
                    seats=seats,
                ),
            )
        )
        return

    if job["kind"] == "booking-cancelled":
        await send_mail(
            Mail(
                to=customer.email,
                subject=f"Cancelled — {booking.reference}",
                html=booking_cancelled_email(
                    reference=booking.reference, event_title=event.title, seats=seats
                ),
            )
        )
        return

    if booking.status != BookingStatus.CONFIRMED:
        print(f"[email] booking {booking.reference} is {booking.status}, not confirming")
        return

    # Decimal arithmetic, never float. A float cannot hold 0.10, and money that
    # is off by a cent in an email is money the customer will ask about.
    total = sum((bs.price_at_booking for bs, _ in rows), Decimal(0))
    qr_data_url = render_qr_data_url(booking.qr_token)

    await send_mail(
        Mail(
            to=customer.email,
            subject=f"Your tickets — {event.title} ({booking.reference})",
            html=booking_confirmed_email(
                reference=booking.reference,
                event_title=event.title,
                venue=venue.name,
                starts_at=_utc_string(show.starts_at),
                seats=seats,
                total=f"{total:.2f}",
                qr_data_url=qr_data_url,
                verify_link=verify_url(booking.qr_token),
            ),
            attachments=[
                {
                    "filename": f"ticket-{booking.reference}.png",
                    "content": list(base64.b64decode(qr_data_url.split(",", 1)[1])),
                }
            ],
        )
    )
    print(f"[email] sent {job['kind']} for {booking.reference} to {customer.email}")


async def _process_offer(entry_id: str) -> None:
    """
    Re-reads the entry rather than trusting the payload.

    By the time a retry runs, the offer may have expired and moved on to
    somebody else — emailing a link that is already dead is worse than emailing
    nothing.
    """
    from ..models import WaitlistEntry

    async with Session() as session:
        entry = (
            (await session.execute(select(WaitlistEntry).where(WaitlistEntry.id == entry_id)))
            .scalars()
            .first()
        )

        if (
            entry is None
            or entry.status != WaitlistStatus.OFFERED
            or not entry.offer_token
            or entry.offer_expires_at is None
        ):
            print(f"[email] waitlist offer {entry_id} is no longer open, skipping")
            return

        customer = (
            (await session.execute(select(User).where(User.id == entry.customer_id)))
            .scalars()
            .one()
        )
        category = (
            (
                await session.execute(
                    select(SeatCategory).where(SeatCategory.id == entry.category_id)
                )
            )
            .scalars()
            .one()
        )
        show = (await session.execute(select(Show).where(Show.id == entry.show_id))).scalars().one()
        event = (
            (await session.execute(select(Event).where(Event.id == show.event_id))).scalars().one()
        )
        venue = (
            (await session.execute(select(Venue).where(Venue.id == event.venue_id))).scalars().one()
        )

        remaining: timedelta = entry.offer_expires_at - utcnow()
        minutes = max(1, round(remaining.total_seconds() / 60))
        token = entry.offer_token

    await send_mail(
        Mail(
            to=customer.email,
            subject=f"A seat opened up — {event.title}",
            html=waitlist_offer_email(
                event_title=event.title,
                venue=venue.name,
                starts_at=_utc_string(show.starts_at),
                category=category.name,
                price=f"{category.price:.2f}",
                minutes=minutes,
                claim_url=offer_url(token),
            ),
        )
    )
    print(f"[email] sent waitlist-offer {entry_id} to {customer.email}")


class WorkerSettings:
    """
    Run with: arq ticket_api.jobs.email_queue.WorkerSettings

    Five tries over roughly a minute and a half. A provider blip should not cost
    somebody their ticket, and a permanent failure should stop quickly.
    """

    functions = [send_email]
    max_tries = 5
    job_timeout = 60
    redis_settings = _redis_settings() or RedisSettings()
